auth/realm/base/src/main/java/org/wildfly/security/auth/realm/JaasSecurityRealm.java

```diff
@@ -103,47 +151,58 @@ public SupportLevel getCredentialAcquireSupport(final Class<? extends Credential
+     * @param callbackHandler callbackHandler to pass to LoginContext
+     * @return the instance of LoginContext
+     * @throws RealmUnavailableException
+     */
+    private LoginContext createLoginContext(final String entry, final Subject subject, final CallbackHandler callbackHandler) throws RealmUnavailableException {
+        if (jaasConfigFilePath != null) {
+            File file = new File(this.jaasConfigFilePath);
+            if (!file.exists() && !file.isDirectory()) {
+                throw ElytronMessages.log.failedToLoadJaasConfigFile();
             }
         }
-        else {
-            try {
-                return new LoginContext(loginConfig, subject, handler);
-            } catch (LoginException le) {
-                throw ElytronMessages.log.failedToCreateLoginContext(le);
+        try {
+            if (jaasConfigFilePath == null) {
+                return new LoginContext(entry, subject, callbackHandler);
+            } else {
+                return new LoginContext(entry, subject, callbackHandler, Configuration.getInstance(DEFAULT_CONFIGURATION_POLICY_TYPE, new URIParameter(jaasConfigFilePath)));
             }
+        } catch (LoginException | NoSuchAlgorithmException le) {
+            throw ElytronMessages.log.failedToCreateLoginContext(le);
         }
     }
 
-    private CallbackHandler createCallbackHandler(final Principal principal, final PasswordGuessEvidence evidence) throws RealmUnavailableException {
-        if (handler == null) {
-            return new DefaultCallbackHandler(principal, evidence);
-        }
-        else {
+    private CallbackHandler createCallbackHandler(final Principal principal, final Evidence evidence) {
+        if (handler != null) {
             try {
-                final CallbackHandler callbackHandler = handler.getClass().newInstance();
-                // preserve backwards compatibility: custom handlers were allowed in the past as long as they had a public setSecurityInfo method.
+                final CallbackHandler callbackHandler = handler.getClass().getConstructor().newInstance();
+                // custom handlers were allowed in the past as long as they had a public setSecurityInfo method. Use this method if it exists
                 final Method setSecurityInfo = handler.getClass().getMethod("setSecurityInfo", Principal.class, Object.class);
                 setSecurityInfo.invoke(callbackHandler, principal, evidence);
                 return callbackHandler;
-            } catch (Exception e) {
-                throw ElytronMessages.log.failedToInstantiateCustomHandler(e);
+            } catch (InstantiationException | IllegalAccessException | NoSuchMethodException | InvocationTargetException e) {
+                // ignore if this method does not exist
+                return handler;
             }
+        } else if (Security.getProperty("auth.login.defaultCallbackHandler") != null) {
```

Comment:
Would be good to create and make use of a constant for "auth.login.defaultCallbackHandler".